### PR TITLE
DBZ-916 Support for tables without primary key

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-916

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/relational/TableSchema.java [hunk 1]
```diff
@@ -108,6 +111,16 @@ public Schema keySchema() {
         return keySchema;
     }
 
+    /**
+     * Get the {@link Schema} that represents the table's primary key or the artificial table's primary key for tables not providing primary key.
+     *
+     * @return the Schema describing the all column's that make up the primary key if table has no explicit primary key
+     */
+    @Override
+    public Schema keySchemaOrProxyKeySchema() {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/TableSchema.java]: Is there a strong reason for exposing this information separately, instead of just via the existing `keySchema()` method? After all, that's what's done for MySQL, too. I.e. why is important for calling code to know about the difference? Or can we get rid of this and just feed the surrogate key definition into the schema when building it?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/TableSchema.java]: @gunnarmorling This was intentional as changing the behaviour of the main method caused issues with existing code. So I opted for incremental approach - provide a new contract for connectors using new framework and keep the behaviour for the existing ones.
When we migrate everything to the new framework the old `keySchema` at all could be removed and the contract simplified.

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/TableSchema.java]: What were these issues? I don't see how this would impact existing code in other connectors: only for SQL Server, the code calling the table schema builder would pass the unique key instead of the PK in this particular case. This should be opaque to any other usages?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/TableSchema.java]: @gunnarmorling SO, the issues on MySQL side identified and fixed. I've simplified the code as well.

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcValueConverters.java [hunk 2]
```diff
@@ -1069,17 +1070,19 @@ else if (data instanceof byte[]) {
                     r.deliver(bytes);
                 }
                 if (byteOrderOfBitType() == ByteOrder.BIG_ENDIAN) {
+                    byte[] bytesNew = Arrays.copyOf(bytes, bytes.length);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcValueConverters.java]: How is this related?

File: debezium-core/src/main/java/io/debezium/relational/TableSchemaBuilder.java [hunk 3]
```diff
@@ -106,13 +108,22 @@ public TableSchema create(String schemaPrefix, String envelopSchemaName, Table t
...
                 addField(valSchemaBuilder, column, mapper);
+                addField(proxyKeySchemaBuilder, column, mapper);
             }
         });
         Schema valSchema = valSchemaBuilder.optional().build();
         Schema keySchema = hasPrimaryKey.get() ? keySchemaBuilder.build() : null;
+        Schema proxyKeySchema = proxyKeySchemaBuilder.build();
 
+        if (keySchema == null) {
+            LOGGER.info("Table '{}' does not have primary key, mapping it to proxy key with schema: {}", tableId, SchemaUtil.asDetailedString(proxyKeySchema));
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/TableSchemaBuilder.java]: I don't think "proxy" is the right name here. Rather "surrogate" or something like that.

**Comment by jpechane** [comment]
@gunnarmorling This decision simplifies things a lot, so I am closing this PR in favour of https://github.com/debezium/debezium/pull/972

**Comment by jpechane** [comment]
> @jpechane, one question about the general approach: IIUC, you use _all_ value columns for the key. That's different from what we do for MySQL or Postgres, where we use the first unique key (MySQL only) or null (Postgres and I reckon MySQL, too, if there's no unique key either).
> 
> As we don't have awareness of unique constraints for SQL Server, I think the strategy should be the same as for PG, i.e. use null. Btw. can you double check on the PG docs: there we speak about unique constraints, too, but IIRC, that's not correct and rather a copy-and-paste glitch from the MySQL docs.
> 
> So bottom line, here's what I think should be the strategy across all the (relational) connectors:
> 
> * Use the PK
> * If there's none, use the first UK, if we know about it (MySQL and probably Oracle)
> * Use null
> 
> Does that sound about right?

@gunnarmorling I've created a follow up issue on this discussion - https://issues.jboss.org/browse/DBZ-1401
